File: clojure/space-age/src/space_age.clj
(ns space-age)

(def period 
  {
    :mercury	0.2408467
    :venus	0.61519726
    :earth	1.0
    :mars	1.8808158
    :jupiter	11.862615
    :saturn	29.447498
    :uranus	84.016846
    :neptune	164.79132
  })

(defn on-planet [planet]
  (with-meta (fn [seconds] (/ seconds 31557600 (period planet))) 
    {:doc (str "Returns someone's age on " 
               (clojure.string/capitalize (name planet)) 
               " based on the given age in seconds")}))

(def on-mercury (on-planet :mercury))
(def on-venus (on-planet :venus))
(def on-earth (on-planet :earth))
(def on-mars (on-planet :mars))
(def on-jupiter (on-planet :jupiter))
(def on-saturn (on-planet :saturn))
(def on-uranus (on-planet :uranus))
(def on-neptune (on-planet :neptune))

File: solutions/elixir/space-age/3/lib/space_age.ex
defmodule SpaceAge do
  @earth_year_in_seconds 31_557_600
  @type planet ::
          :mercury
          | :venus
          | :earth
          | :mars
          | :jupiter
          | :saturn
          | :uranus
          | :neptune

  @doc """
  Return the number of years a person that has lived for 'seconds' seconds is
  aged on 'planet', or an error if 'planet' is not a planet.
  """
  @spec age_on(planet, pos_integer) :: {:ok, float} | {:error, String.t()}
  def age_on(:mercury, seconds), do: {:ok, seconds / @earth_year_in_seconds / 0.2408467}
  def age_on(:venus, seconds), do: {:ok, seconds / @earth_year_in_seconds / 0.61519726}
  def age_on(:earth, seconds), do: {:ok, seconds / @earth_year_in_seconds / 1.0}
  def age_on(:mars, seconds), do: {:ok, seconds / @earth_year_in_seconds / 1.8808158}
  def age_on(:jupiter, seconds), do: {:ok, seconds / @earth_year_in_seconds / 11.862615}
  def age_on(:saturn, seconds), do: {:ok, seconds / @earth_year_in_seconds / 29.447498}
  def age_on(:uranus, seconds), do: {:ok, seconds / @earth_year_in_seconds / 84.016846}
  def age_on(:neptune, seconds), do: {:ok, seconds / @earth_year_in_seconds / 164.79132}
  def age_on(_, _), do: {:error, "not a planet"}
end
